Framework/API/test/MatrixWorkspaceTest.h

```diff
@@ -1291,6 +1291,97 @@ class MatrixWorkspaceTest : public CxxTest::TestSuite {
+    TSM_ASSERT_EQUALS("Should have a length of 3", ws.dataDx(0).size(), j);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.dataDx(0)[0],
+                      value0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.dataDx(0)[1],
+                      value0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.dataDx(0)[2],
+                      value0);
+
+    TSM_ASSERT("Should have x resolution values", ws.hasDx(1));
+    TSM_ASSERT_EQUALS("Should have a length of 3", ws.dataDx(1).size(), j);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.dataDx(1)[0],
+                      value1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.dataDx(1)[1],
+                      value1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.dataDx(1)[2],
+                      value1);
+
+    TSM_ASSERT("Should have x resolution values", ws.hasDx(2));
+    TSM_ASSERT_EQUALS("Should have a length of 3", ws.dataDx(2).size(), j);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.dataDx(2)[0],
+                      value2);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.dataDx(2)[1],
+                      value2);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.dataDx(2)[2],
+                      value2);
+
+    // Test readDx
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.readDx(0)[0],
+                      value0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.readDx(0)[1],
+                      value0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", ws.readDx(0)[2],
+                      value0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.readDx(1)[0],
+                      value1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.readDx(1)[1],
+                      value1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", ws.readDx(1)[2],
+                      value1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.readDx(2)[0],
+                      value2);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.readDx(2)[1],
+                      value2);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", ws.readDx(2)[2],
+                      value2);
+
+    // RefY
+    auto ref0 = ws.refDx(0);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 10", (*ref0)[0],
+                      value0);
+    auto ref1 = ws.refDx(1);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 11", (*ref1)[1],
+                      value1);
+    auto ref2 = ws.refDx(2);
+    TSM_ASSERT_EQUALS("Should have x resolution values of 12", (*ref2)[2],
+                      value2);
+
+    TSM_ASSERT("Should not have any x resolution values", !ws.hasDx(3));
+  }
+
 private:
```

Comment:
While you're modifying this test, would you also cleanup the member function below? 

- use `boost::make_shared` instead of `new`
- Change the outer loop to a range-based for loop.
- The inner loop can be replaced by [std::iota](http://www.cplusplus.com/reference/numeric/iota/).